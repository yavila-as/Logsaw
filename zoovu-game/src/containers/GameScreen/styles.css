.imageBox {
    background-color: white;
    width: 60px;
    height: 60px;
    border-radius: 10px;
    padding: 20px;
    margin: 10px;
}

.borderBox {
    width: 100px;
    height: 100px;
    border-radius: 10px;
    margin: 10px;
}