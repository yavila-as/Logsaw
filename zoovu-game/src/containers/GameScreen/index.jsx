import React, { Component } from 'react'
import Box from '@material-ui/core/Button';
import { Grid, Typography } from '@material-ui/core';
import imageZ from '../../images/zoovuz.svg';
import imageo from '../../images/zoovuo.svg';
import imagev from '../../images/zoovuv.svg';
import imageu from '../../images/zoovuu.svg';
import boxImage from '../../images/greenBox.png';
import './styles.css';

const initialState = {
  list: [
      { number: 1, image: `${imageZ}`},
      { number: 2, image: `${imageo}`},
      { number: 2, image: `${imageo}`},
      { number: 3, image: `${imagev}`},
      { number: 4, image: `${imageu}`},
  ],
  editableList: [
      { number: 5, image: `${boxImage}`},
      { number: 6, image: `${boxImage}`},
      { number: 7, image: `${boxImage}`},
      { number: 8, image: `${boxImage}`},
      { number: 9, image: `${boxImage}`},
  ],
  shuffledList: [],
  seconds: 0,
  gameStarted: false,
  win: false,
};

export default class GameScreen extends Component {
  constructor(props) {
    super(props)
    this.state = initialState;
  }

  componentDidMount(){
    this.shufflePieces();
  }
  
  shufflePieces = () =>{
    const shuffled = this.state.list;

    for (let i = shuffled.length - 1; i > 0; i--) {
      let j = Math.floor(Math.random() * (i + 1));
      let tmp = shuffled[i];
      shuffled[i] = shuffled[j];
      shuffled[j] = tmp;
    }
    this.setState({shuffledList: shuffled});
  }

  timeCounter = (time) => {
    this.timer = setInterval(() => {
      this.setState({
        seconds : this.state.seconds + time
      })
    }, 1000)
  }

  onDragStart = (event) => {
    if (!this.state.gameStarted) {
      this.timeCounter(1);
      this.setState({ gameStarted: true });
    } 
    const initialPosition = Number(event.currentTarget.dataset.position);
    event.dataTransfer.setData("draggedFrom", initialPosition);
  }
  
  onDragOver = (event) => {
    event.preventDefault();
  }

  onDrop = (event) => {
    let draggedFrom = event.dataTransfer.getData("draggedFrom");
    let draggedTo = Number(event.currentTarget.dataset.position);
    //if(this.state.editableList[draggedTo].image === `${boxImage}`){
      const itemDragged = this.state.shuffledList[draggedFrom];
      const remainingItems = this.state.editableList.filter((item, index) => index !== Number(draggedTo));
  
      let editableList = [
        ...remainingItems.slice(0, draggedTo),
        itemDragged,
        ...remainingItems.slice(draggedTo)
      ];
      const itemDraggedA = this.state.editableList[draggedTo];
      const remainingItemsA = this.state.shuffledList.filter((item, index) => index !== Number(draggedFrom));
      let shuffledList = [
        ...remainingItemsA.slice(0, draggedFrom),
        itemDraggedA,
        ...remainingItemsA.slice(draggedFrom)
      ];
      this.setState({
        ...this.state,
        editableList,
        shuffledList
      });
      const temp = [1, 2, 2, 3, 4];
      if(this.state.shuffledList[draggedFrom].number !== temp[draggedTo]){
        this.setState({
          seconds : this.state.seconds + 10
        })
      }
      const listChecked = editableList.map((element, index) => {
        return element.number === temp[index] ? true : false;
      });
      this.setState({
        win: !listChecked.includes(false) && !editableList.some(e => e.image === `${boxImage}`),
      }, () =>{
        if(this.state.win){
          clearInterval(this.timer);
          setTimeout(()=> this.onUserWin(), 10000);
        }
      });
   // }
  }
   
  onUserWin(){
    alert('You won!!! congratulations');
    this.setState(initialState);
    this.shufflePieces();
  }

  render() {
    return (
      <Box display='flex' flexWrap="wrap">
        <Grid
            container
            direction="column"
            justify="left"
            alignItems="left"
        >
          <Typography>Good luck,{this.props.userName}!</Typography>
          <p>
              Your score: {this.state.seconds} seconds
          </p>
          <Typography>Pick up the right cards</Typography>
          <section>
              {this.state.shuffledList.map( (item, index) => (
                  <item 
                  key={index}
                  data-position={index}
                  draggable
                  onDragStart={(event) => this.onDragStart(event)}
                  onDragOver={(event) => this.onDragOver(event)}
                  onDrop={(event) => this.onDrop(event)}
                  >
                      <img src={item.image} className="imageBox"/>
                  </item>
              ))}
              </section>
          <Typography>..and drop them here to make the logo great again!</Typography>
          <section>
              {this.state.editableList.map( (item, index) => {
              return(
              <item 
              key={index}
              data-position={index}
              draggable
              onDragStart={(event) => this.onDragStart(event)}
              onDragOver={(event) => this.onDragOver(event)}
              onDrop={(event) => this.onDrop(event)}
              >
                  <img src={item.image} className="imageBox"/>
              </item>
              )
              })}
          </section>
        </Grid>
      </Box>
    );
  }
}
