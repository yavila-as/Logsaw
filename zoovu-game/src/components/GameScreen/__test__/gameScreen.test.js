import React from 'react';
import ReactDOM from 'react-dom';
import GameScreen from '../index';
import { render, cleanup } from '@testing-library/react';
import "@testing-library/jest-dom/extend-expect";

import renderer from 'react-test-renderer';

afterEach(cleanup);

it("matches snapshot", () =>{
   const tree = renderer.create(<GameScreen></GameScreen>).toJSON();
   expect(tree).toMatchSnapshot();
})