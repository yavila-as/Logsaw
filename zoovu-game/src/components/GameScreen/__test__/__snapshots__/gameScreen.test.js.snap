// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text"
  disabled={false}
  display="flex"
  flexwrap="wrap"
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 MuiTypography-alignLeft"
      >
        Good luck, 
        !
      </p>
      <p>
        Your score: 
        0
         seconds
      </p>
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        Pick up the right cards
      </p>
      <section>
        <item
          data-position={0}
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
        >
          <img
            className="imageBox"
            src="zoovuz.svg"
          />
        </item>
        <item
          data-position={1}
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
        >
          <img
            className="imageBox"
            src="zoovuo.svg"
          />
        </item>
        <item
          data-position={2}
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
        >
          <img
            className="imageBox"
            src="zoovuu.svg"
          />
        </item>
        <item
          data-position={3}
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
        >
          <img
            className="imageBox"
            src="zoovuo.svg"
          />
        </item>
        <item
          data-position={4}
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
        >
          <img
            className="imageBox"
            src="zoovuv.svg"
          />
        </item>
      </section>
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        ..and drop them here to make the logo great again!
      </p>
      <section>
        <item
          data-position={0}
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
        >
          <img
            className="imageBox"
            src="greenBox.png"
          />
        </item>
        <item
          data-position={1}
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
        >
          <img
            className="imageBox"
            src="greenBox.png"
          />
        </item>
        <item
          data-position={2}
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
        >
          <img
            className="imageBox"
            src="greenBox.png"
          />
        </item>
        <item
          data-position={3}
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
        >
          <img
            className="imageBox"
            src="greenBox.png"
          />
        </item>
        <item
          data-position={4}
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
        >
          <img
            className="imageBox"
            src="greenBox.png"
          />
        </item>
      </section>
    </div>
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</button>
`;
